src/actions/copy_object.rs

```diff
@@ -0,0 +1,151 @@
+use std::time::Duration;
+
+use time::OffsetDateTime;
+use url::Url;
+
+use super::S3Action;
+use crate::actions::Method;
+use crate::signing::sign;
+use crate::{Bucket, Credentials, Map};
+
+/// Create a copy of an object that is already stored in S3, using a `PUT` request.
+///
+/// Find out more about CopyObject from the [AWS API Reference][api]
+///
+/// [api]: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html
+#[derive(Debug, Clone)]
+pub struct CopyObject<'a> {
+    bucket: &'a Bucket,
+    credentials: Option<&'a Credentials>,
+    src_object: &'a str,
+    dst_object: &'a str,
+
+    query: Map<'a>,
+    headers: Map<'a>,
+}
+
+impl<'a> CopyObject<'a> {
+    #[inline]
+    pub fn new(
+        bucket: &'a Bucket,
+        credentials: Option<&'a Credentials>,
+        src_object: &'a str,
+        dst_object: &'a str,
+    ) -> Self {
+        let mut query = Map::new();
+        query.insert(
+            "x-amz-copy-source",
+            format!("{}/{}", bucket.name(), src_object),
+        );
+
+        Self {
+            bucket,
+            credentials,
+            src_object,
+            dst_object,
+
+            query,
+            headers: Map::new(),
+        }
+    }
+}
+
+impl<'a> S3Action<'a> for CopyObject<'a> {
+    const METHOD: Method = Method::Put;
+
+    fn sign(&self, expires_in: Duration) -> Url {
+        let now = OffsetDateTime::now_utc();
+        self.sign_with_time(expires_in, &now)
+    }
+
+    fn query_mut(&mut self) -> &mut Map<'a> {
+        &mut self.query
+    }
+
+    fn headers_mut(&mut self) -> &mut Map<'a> {
+        &mut self.headers
+    }
+
+    fn sign_with_time(&self, expires_in: Duration, time: &OffsetDateTime) -> Url {
+        let mut url = self.bucket.object_url(self.dst_object).unwrap();
+
+        match self.credentials {
+            Some(credentials) => sign(
+                time,
+                Method::Put,
+                url,
+                credentials.key(),
+                credentials.secret(),
+                credentials.token(),
+                self.bucket.region(),
+                expires_in.as_secs(),
+                self.query.iter(),
+                self.headers.iter(),
+            ),
+            None => {
+                url.query_pairs_mut().append_pair(
```

Comment:
This should instead call `crate::signing::util::add_query_params(url, self.query.iter())` so that any custom query parameters are also included.

Example from another action: https://github.com/paolobarbolini/rusty-s3/blob/9e10f5f9265d4bad649b473c4ca7c49b419dab50/src/actions/list_objects_v2.rs#L149